activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -442,10 +442,21 @@ def using_limitable_reflections?(reflections)
       def find_with_ids(*ids)
         raise UnknownPrimaryKey.new(@klass) if primary_key.nil?
 
-        expects_array = ids.first.kind_of?(Array)
+        expects_array = if klass.composite_primary_key?
+          ids.first.first.is_a?(Array)
+        else
+          ids.first.is_a?(Array)
+        end
+
         return [] if expects_array && ids.first.empty?
 
-        ids = ids.flatten.compact.uniq
+        ids = if klass.composite_primary_key?
+          expects_array ? ids.first : ids
+        else
+          ids.flatten
+        end
```

Comment:
This feels hard to follow. The flatten made sense before as an easy way of ignoring `expects_array`... but now I think we can just always respect that:

```suggestion
        ids = expects_array ? ids.first : ids
```

Reply:
Totally agree, the reason I decided to keep `flatten` for a regular model as it was handling cases like `MyNonCpkModel.find([1,[1,2],[3,4,5],[[[6]]]])` which doesn't look like a supported API though it was working and I decided to keep it to avoid breaking some use-case which I'm not aware of.

But if we all agree that `find` does not support passing non-flattened arrays then `ids.first` will work for all models. I'll make the change